crates/store/src/db/sql/mod.rs

```diff
@@ -190,6 +193,317 @@ pub fn select_accounts_by_ids(
+    )?;
+
+    let mut select_latest_non_fungible_assets_stmt = conn.prepare_cached(
+        "
+        SELECT
+            account_id, vault_key
+        FROM
+            account_non_fungible_asset_updates a
+        WHERE
+            account_id IN rarray(?1) AND
+            block_num <= ?2 AND
+            is_remove = 0 AND
+            NOT EXISTS(
+                SELECT 1
+                FROM account_non_fungible_asset_updates b
+                WHERE
+                    a.account_id = b.account_id AND
+                    b.is_remove = 1 AND
+                    b.block_num <= ?2 AND
+                    a.vault_key = b.vault_key AND
+                    a.block_num < b.block_num
+            )
+        ORDER BY
+            account_id
+        ",
+    )?;
+
+    struct AccountRecord {
+        code: AccountCode,
+        nonce: Option<Felt>,
+        storage: BTreeMap<u8, StorageSlot>,
+        assets: Vec<Asset>,
+    }
+
+    let account_ids = Rc::new(account_ids.iter().copied().map(u64_to_value).collect::<Vec<_>>());
+
+    // Gathering data from different tables to single accounts map.
+    let mut accounts = BTreeMap::new();
+
+    let mut rows = select_last_states_stmt.query(params![Rc::clone(&account_ids)])?;
+    while let Some(row) = rows.next()? {
+        let account_id: AccountId = row.get(0)?;
+        let account_details = row.get_ref(1)?.as_blob_or_null()?;
+        if let Some(details) = account_details {
+            let details = Account::read_from_bytes(details)?;
+            accounts.insert(
+                account_id,
+                AccountRecord {
+                    code: details.code().clone(),
+                    nonce: None,
+                    storage: BTreeMap::new(),
+                    assets: vec![],
+                },
+            );
+        }
+    }
+
+    let mut rows =
+        select_latest_nonces_stmt.query(params![Rc::clone(&account_ids), block_number])?;
+    while let Some(row) = rows.next()? {
+        let account_id: AccountId = row.get(0)?;
+        let nonce: u64 = row.get(1)?;
+        find_account(&mut accounts, account_id)?.get_mut().nonce =
+            Some(nonce.try_into().map_err(DatabaseError::DataCorrupted)?);
+    }
+
+    let mut rows = select_latest_storage_slot_updates_stmt
+        .query(params![Rc::clone(&account_ids), block_number])?;
+    while let Some(row) = rows.next()? {
+        let account_id: AccountId = row.get(0)?;
+        let slot: u8 = row.get(1)?;
+        let value = row.get_ref(2)?.as_blob()?;
+        let value = Word::read_from_bytes(value)?;
+
+        match find_account(&mut accounts, account_id)?.get_mut().storage.entry(slot) {
+            Entry::Vacant(entry) => {
+                entry.insert(StorageSlot::Value(value));
+            },
+            Entry::Occupied(_) => {
+                return Err(DatabaseError::DataCorrupted(format!(
+                    "Duplicate storage slot: {slot}"
+                )));
+            },
+        }
```

Comment:
Could use `if <..>.insert(..).is_some() { return Err(..) }`. Unclear if its more legible though.

```suggestion
        let mut account = find_account(&mut accounts, account_id)?;
        if account.get_mut().storage.insert(slot, value).is_some() {
            return Err(DatabaseError::DataCorrupted(format!(
                "Duplicate storage slot: {slot}"
            )));
        }
```